### PR TITLE
Cams 84 assign trial attorney 2

### PR DESCRIPTION
**Pull request by Swathi-Darbha**
# Purpose

Assign a Trial Attorney to a Chapter 15 Case in the CAMS application backend.

# Major Changes

An Azure Function with an HttpTrigger has been added to respond to an HTTP Post request via the URL : _http://server:port/api/case-assignments_

A CaseAssignment Controller, related Use Case, domain entity for the Case Assignment, request and response models, as well as the local repository have been added, which all handle the case assignment workflow.

# Testing/Validation 
Steps to test the case assignment to a trial attorney using postman: 

1. Start and run the backend NodeJs application in the given environment.
2. In order to create an assignment for one or more Trial Attorneys to a given case, the _/case-assignments_ URL accepts the following input parameters in an http post request:

-   CaseId, 
-   A collection of Attorney Ids,
-   A role called 'TrialAttroney' 

3. Open Postman.
4. In the drop down for Request type , choose POST.
5. In the URL field, enter the URL of the backend application with the server and port number for your environment. E.g. 
           http://servername:port/api/case-assignments.
6. Go to the tab 'Body' . Select the radio button option for 'raw'. Then paste the request in a json format in the below text 
     box. Sample request below:
    
    ``` 
    {
        "caseId" : 123,
        "attorneyIdList": [345, 678],
        "role" : "TrialAttorney"
    }

7.  Click on the 'Send' button, which runs the POST request for case-assignments.



Functionality : An http Response is returned only if the input parameters pass all the validation tests in the azure function.
Once the validation is done, an assignment is created in the backend for each of the attorneyIds passed and the appropriate response is returned with a success 200K message along with the AssignmentId created in the backend. 

Test Error handling: Run the URL without passing one of the input parameters above and an error message is sent back with an http bad Request 400.

# Notes

This current PR creates a case assignment in the local repository.
Future stories/tasks will add the connections to the CAMS Cosmos DB.

### CODE REVIEW

File: backend/functions/caseassignments/case.assignment.function.test.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    };
+    await httpTrigger(appContext, request);
+    expect(appContext.res.body).toEqual(expectedResponse);
+  });
+
+  test('returns response with multiple assignment Ids , when requested to create assignments for multiple trial attorneys on a case', async () => {
+    const request = {
+      query: {
+        caseId: '6789',
+        attorneyIdList: ['2082', '2083', '2082'],
```
Comment (Swathi-Darbha) [inline backend/functions/caseassignments/case.assignment.function.test.ts:32]: Handle duplicate attoney Ids in a separate test.

File: backend/functions/caseassignments/case.assignment.function.test.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        caseId: '6789',
+        attorneyIdList: ['2082'],
+        role: 'TrialAttorney',
+      },
+    };
+
+    const httpErrorSpy = jest.spyOn(httpModule, 'httpError');
+    await httpTrigger(context, request);
+
+    expect(httpErrorSpy).toHaveBeenCalled();
```
Comment (Swathi-Darbha) [inline backend/functions/caseassignments/case.assignment.function.test.ts:140]: Add a check to verify the Error Message as well.

File: backend/functions/lib/interfaces/ICaseAssignmentRepository.d.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { ApplicationContext } from '../adapters/types/basic';
+import { CaseAttorneyAssignment } from '../adapters/types/case.attorney.assignment';
+
+export interface ICaseAssignmentRepository {
```
Comment (Swathi-Darbha) [inline backend/functions/lib/interfaces/ICaseAssignmentRepository.d.ts:4]: modify the interface naming : append the word interface at the end. Remove the I at the beginning.

File: backend/functions/lib/interfaces/chapter-15-data-interface.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
-import { Chapter15Case } from '../adapters/types/cases';
+import { IChapter15Case } from '../adapters/types/cases';
 import { PacerCaseData } from '../adapters/types/pacer';
 
-export function pacerToChapter15Data(input: PacerCaseData[]): Chapter15Case[] {
+export function pacerToChapter15Data(input: PacerCaseData[]): IChapter15Case[] {
```
Comment (Swathi-Darbha) [inline backend/functions/lib/interfaces/chapter-15-data-interface.ts:4]: Update the interface name here and elsewhere in the code.

File: backend/functions/lib/adapters/types/trial.attorney.assignment.response.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export class TrialAttorneyAssignmentResponse {
```
Comment (Swathi-Darbha) [inline backend/functions/lib/adapters/types/trial.attorney.assignment.response.ts:1]: Modify this to a type interface and move into a types file. Similar to cases.d.ts and similar variable naming.

File: backend/functions/lib/use-cases/chapter-15.case.assignment.test.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+      resultAssignmentId1,
+    );
+    const assignmentCreated2 = await mockCaseAssignmentRepository.getAssignment(
+      resultAssignmentId2,
+    );
+    expect(resultAssignmentId2).toBe(resultAssignmentId1);
+    expect(assignmentCreated2).toEqual(assignmentCreated1);
+  });
+
+  test('create a new trial attorney assignment on an existing case succeeds', async () => {
```
Comment (Swathi-Darbha) [inline backend/functions/lib/use-cases/chapter-15.case.assignment.test.ts:80]: This test should not succeed. Cannot do a new assignment on an existing case.

File: backend/functions/lib/use-cases/chapter-15.case.assignment.test.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
+import { CaseAssignmentRole } from '../adapters/types/case.assignment.role';
+import { CaseAssignmentController } from '../adapters/controllers/case.assignment.controller';
```
Comment (Swathi-Darbha) [inline backend/functions/lib/use-cases/chapter-15.case.assignment.test.ts:2]: Since these are testing the controller, this file needs to move to the controllers folder.

File: backend/functions/lib/adapters/gateways/case.assignment.local.repository.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+export class CaseAssignmentLocalRepository implements ICaseAssignmentRepository {
+  private caseAttorneyAssignments: CaseAttorneyAssignment[] = [];
+  private _nextUnusedId = 1;
+
+  public async createAssignment(
+    context: ApplicationContext,
+    caseAssignment: CaseAttorneyAssignment,
+  ): Promise<number> {
+    const assignment: CaseAttorneyAssignment = await this.findAssignment(caseAssignment);
```
Comment (Swathi-Darbha) [inline backend/functions/lib/adapters/gateways/case.assignment.local.repository.ts:13]: If an assignment has already been created on a case, and another assignment request for the same case comes in, respond with a 400 bad request.

File: backend/functions/caseassignments/case.assignment.function.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+const REQUIRED_PROFESSIONAL_ID_MESSAGE = 'Required parameter professionalId is absent.';
+const REQUIRED_ROLE_MESSAGE = 'Required parameter - role of the attorney is absent.';
+const INVALID_ROLE_MESSAGE =
+  'Invalid role for the professional. Requires role to be a TrialAttorney for case assignment';
+
+const httpTrigger: AzureFunction = async function (
+  functionContext: Context,
+  request: HttpRequest,
+): Promise<void> {
+  const caseId = request.query.caseId || (request.body && request.body.caseId);
```
Comment (Swathi-Darbha) [inline backend/functions/caseassignments/case.assignment.function.ts:21]: Allow the request to be passed only in the body and not the query.

File: backend/functions/lib/use-cases/case.assignment.service.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+  constructor(assignmentRepository?: ICaseAssignmentRepository) {
+    if (!assignmentRepository) {
+      this._assignmentRepository = getAssignmentRepository();
+    } else {
+      this._assignmentRepository = assignmentRepository;
+    }
+  }
+
+  async createAssignment(
```
Comment (Swathi-Darbha) [inline backend/functions/lib/use-cases/case.assignment.service.ts:18]: Add functionality for the following in the business use case:
If an assignment has already been created on a case, and another assignment request for the same case comes in, respond with a 400 bad request.

File: backend/functions/lib/adapters/types/case.attorney.assignment.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { CaseAssignmentRole } from './case.assignment.role';
+
+export class CaseAttorneyAssignment {
+  _assignmentId: number;
+  _caseId: string;
+  _caseTitle: string;
+  _professionalId: string;
+  _role: CaseAssignmentRole;
```
Comment (jamesobrooks) [inline backend/functions/lib/adapters/types/case.attorney.assignment.ts:8]: These are not private and you are accessing them in another test.

File: backend/functions/caseassignments/case.assignment.function.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+import { CaseAssignmentRole } from '../lib/adapters/types/case.assignment.role';
+import { TrialAttorneysAssignmentRequest } from '../lib/adapters/types/trial.attorneys.assignment.request';
+import { CaseAssignmentException } from './case.assignment.exception';
+
+const NAMESPACE = 'CASE-ASSIGNMENT-FUNCTION' as const;
+const REQUIRED_CASE_ID_MESSAGE = 'Required parameter caseId is absent.';
+const REQUIRED_PROFESSIONAL_ID_MESSAGE = 'Required parameter professionalId is absent.';
+const REQUIRED_ROLE_MESSAGE = 'Required parameter - role of the attorney is absent.';
+const INVALID_ROLE_MESSAGE =
+  'Invalid role for the professional. Requires role to be a TrialAttorney for case assignment';
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.ts]: Done.

File: backend/functions/lib/use-cases/case.assignment.service.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
+import { ICaseAssignmentRepository } from '../interfaces/ICaseAssignmentRepository';
+import { getAssignmentRepository } from '../factory';
+import { CaseAttorneyAssignment } from '../adapters/types/case.attorney.assignment';
+import { ApplicationContext } from '../adapters/types/basic';
+import { TrialAttorneyAssignmentResponse } from '../adapters/types/trial.attorney.assignment.response';
+
+export class CaseAssignmentService {
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/case.assignment.service.ts]: Done.

File: backend/functions/lib/interfaces/ICaseAssignmentRepository.d.ts (hunk 3 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/interfaces/ICaseAssignmentRepository.d.ts]: Done.

File: backend/functions/lib/adapters/types/trial.attorney.assignment.response.ts (hunk 5 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/types/trial.attorney.assignment.response.ts]: Done.

File: backend/functions/lib/interfaces/chapter-15-data-interface.ts (hunk 4 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/interfaces/chapter-15-data-interface.ts]: Done.

File: backend/functions/lib/use-cases/chapter-15.case.assignment.test.ts (hunk 7 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/chapter-15.case.assignment.test.ts]: Done.

File: backend/functions/caseassignments/case.assignment.function.test.ts (hunk 2 above)
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.test.ts:140]: Done.

File: backend/functions/caseassignments/case.assignment.function.ts (hunk 9 above)
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.ts]: Done.

File: backend/functions/caseassignments/case.assignment.function.test.ts (hunk 1 above)
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.test.ts:34]: Done.

File: backend/functions/lib/adapters/types/case.attorney.assignment.ts (hunk 11 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/types/case.attorney.assignment.ts]: Done, updated the naming on the class variables, based on our meeting discussion to use the google guidelines going forward.

File: backend/functions/caseassignments/case.assignment.function.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  role: string,
+) {
+  validateCaseId(caseId, functionContext);
+  validateProfessionalId(listOfAttorneyIds, functionContext);
+  validateRole(role, functionContext);
+}
+
+function validateCaseId(caseId: string, functionContext: Context) {
+  if (!caseId) {
+    log.error(applicationContextCreator(functionContext), NAMESPACE, REQUIRED_CASE_ID_MESSAGE);
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.ts:75]: Added Logging. Done.

File: backend/functions/lib/use-cases/chapter-15.case.assignment.test.ts (hunk 6 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/chapter-15.case.assignment.test.ts]: Fixed the test to fail. Done.

File: backend/functions/lib/use-cases/case.assignment.service.ts (hunk 10 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/case.assignment.service.ts]: Done. Added functionality to check if an assignment already exists and throw an error in this scenario.

File: backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+      const assignmentResponse2 = await assignmentController.createTrailAttorneyAssignments(
+        testCaseAssignment2,
+      );
+
+      resultAssignmentId2 = assignmentResponse2.assignmentIdList[0];
+    } catch (exception) {
+      if (exception instanceof AssignmentException && exception.status === 400) {
+        context.res = httpError(context, exception, 400);
+      } else {
+        context.res = httpError(context, exception, 500);
```
Comment (jamesobrooks) [inline backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts:127]: Why are we setting this up in a test?

File: backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+    const expectedNumberOfAssignments: number = 1;
+    const actualNumberOfAssignments = await mockCaseAssignmentRepository.getCount();
+    expect(actualNumberOfAssignments).toBe(expectedNumberOfAssignments);
+    expect(assignmentCreated1.caseId).toBe(testCaseAssignment1.caseId);
+    expect(assignmentCreated1.attorneyId).toBe(testCaseAssignment1.listOfAttorneyIds[0]);
+    expect(assignmentCreated1.role).toBe(testCaseAssignment1.role);
+    expect(assignmentCreated2).toBeFalsy();
+    expect(context.res.body).toEqual(expectedResponse);
+    expect(context.res.statusCode).toBe(400);
+    expect(context.res.body);
```
Comment (jamesobrooks) [inline backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts:148]: Why are we looking at function invocation context when testing the controller?

File: backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+    expect(assignmentCreated1.caseId).toBe(testCaseAssignment.caseId);
+    expect(assignmentCreated1.attorneyId).toBe(testCaseAssignment.listOfAttorneyIds[0]);
+    expect(assignmentCreated1.role).toBe(testCaseAssignment.role);
+
+    const resultAssignmentId2 = assignmentResponse.assignmentIdList[1];
+    const assignmentCreated2 = await mockCaseAssignmentRepository.getAssignment(
+      resultAssignmentId2,
+    );
+
+    expect(resultAssignmentId2).toBeGreaterThan(0);
```
Comment (jamesobrooks) [inline backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts:192]: I'm not sure this is a meaningful test. Is there nothing more we can know about the id than that it is greater than 0? Are you able to break this test (i.e. can you make the id less than 0)?

File: backend/functions/lib/adapters/gateways/case.assignment.local.repository.ts [hunk 18]
```diff
@@ -0,0 +1,53 @@
+import { CaseAssignmentRepositoryInterface } from '../../interfaces/case.assignment.repository.interface';
+import { CaseAttorneyAssignment } from '../types/case.attorney.assignment';
+import { ApplicationContext } from '../types/basic';
+
+export class CaseAssignmentLocalRepository implements CaseAssignmentRepositoryInterface {
+  private caseAttorneyAssignments: CaseAttorneyAssignment[] = [];
+  private _nextUnusedId = 1;
```
Comment (jamesobrooks) [inline backend/functions/lib/adapters/gateways/case.assignment.local.repository.ts:7]: We agreed to use Google's TypeScript style guide starting with this change and to leave the addition to documentation for a separate pr.

File: backend/functions/lib/adapters/types/case.assignment.request.ts [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+  set caseTitle(value: string) {
+    this._caseTitle = value;
+  }
+  get caseId(): string {
+    return this._caseId;
+  }
+
+  set caseId(value: string) {
+    this._caseId = value;
+  }
```
Comment (jamesobrooks) [inline backend/functions/lib/adapters/types/case.assignment.request.ts:67]: These accessors only act as passthroughs, so I don't think it is meaningful for the properties to be private.

**Review by jamesobrooks — CHANGES_REQUESTED** [review]

File: backend/functions/caseassignments/case.assignment.function.test.ts [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+      },
+    };
+    const expectedResponse = {
+      error: 'Required parameter caseId is absent.',
+    };
+
+    await httpTrigger(appContext, request);
+    expect(appContext.res.body).toEqual(expectedResponse);
+    expect(appContext.res.statusCode).toBe(400);
+    expect(appContext.res.body);
```
Comment (jamesobrooks) [inline backend/functions/caseassignments/case.assignment.function.test.ts:86]: I don't believe we need this (or the similar lines in a couple of the other tests).

File: backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+describe('Chapter 15 Case Assignment Creation Tests', () => {
+  test('A chapter 15 case is assigned to an attorney when requested', async () => {
+    const testCaseAssignment = new TrialAttorneysAssignmentRequest(
+      '12345',
+      ['8082'],
+      CaseAssignmentRole.TrialAttorney,
+    );
+
+    const mockCaseAssignmentRepository: CaseAssignmentRepositoryInterface =
+      new CaseAssignmentLocalRepository();
```
Comment (jamesobrooks) [inline backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts:19]: These mock repositories are not actually mocks but real instances of the local repository.

File: backend/functions/caseassignments/case.assignment.function.ts [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+    const trialAttorneyAssignmentResponse =
+      await caseAssignmentController.createTrailAttorneyAssignments(assignmentRequest);
+    functionContext.res = httpSuccess(functionContext, trialAttorneyAssignmentResponse);
+    console.log(functionContext.res.body.toString());
+  } catch (exception) {
+    if (exception instanceof CaseAssignmentException && exception.status === 400) {
+      functionContext.res = httpError(functionContext, exception, 400);
+    } else {
+      functionContext.res = httpError(functionContext, exception, 500);
+    }
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.ts:45]: Done.

File: backend/functions/caseassignments/case.assignment.function.ts [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+) {
+  validateCaseId(caseId, functionContext);
+  validateProfessionalId(listOfAttorneyIds, functionContext);
+  validateRole(role, functionContext);
+}
+
+function validateCaseId(caseId: string, functionContext: Context) {
+  if (!caseId) {
+    log.error(applicationContextCreator(functionContext), NAMESPACE, REQUIRED_CASE_ID_MESSAGE);
+    functionContext.res = httpError(functionContext, new Error(REQUIRED_CASE_ID_MESSAGE), 400);
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.ts:76]: Done. Removed redundant writing to the context response.

File: backend/functions/caseassignments/case.assignment.function.ts [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    throw new CaseAssignmentException(400, REQUIRED_CASE_ID_MESSAGE);
+  }
+}
+
+function validateProfessionalId(listOfAttorneyIds: string[], functionContext: Context) {
+  if (listOfAttorneyIds.length < 1) {
+    log.error(applicationContextCreator(functionContext), NAMESPACE, REQUIRED_ATTORNEY_ID_MESSAGE);
+    functionContext.res = httpError(functionContext, new Error(REQUIRED_ATTORNEY_ID_MESSAGE), 400);
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.ts:84]: Done. Removed redundant writing to the context response.

File: backend/functions/caseassignments/case.assignment.function.ts [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+    throw new CaseAssignmentException(400, REQUIRED_ATTORNEY_ID_MESSAGE);
+  }
+}
+
+function validateRole(role: string, functionContext: Context) {
+  if (!role) {
+    log.error(applicationContextCreator(functionContext), NAMESPACE, REQUIRED_ROLE_MESSAGE);
+    functionContext.res = httpError(functionContext, new Error(REQUIRED_ROLE_MESSAGE), 400);
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.ts:92]: Done. Removed redundant writing to the context response.

File: backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts (hunk 15 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts]: Done, fixed.

File: backend/functions/lib/factory.ts [hunk 26]
```diff
@@ -43,3 +45,12 @@ export const getPacerGateway = (): CasesInterface => {
 export const getPacerTokenSecretGateway = (): PacerSecretsInterface => {
   return new PacerSecretsGateway();
 };
+
+export const getAssignmentRepository = (): CaseAssignmentRepositoryInterface => {
+  const config: ApplicationConfiguration = new ApplicationConfiguration();
+  if (config.get('dbMock')) {
+    return new CaseAssignmentLocalRepository();
+  } else {
+    // return new CaseAssignmentRepository(); // not yet implemented
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/factory.ts:54]: Done.

File: backend/functions/lib/use-cases/assignment.exception.ts [hunk 27]
```diff
@@ -0,0 +1,9 @@
+export class AssignmentException extends Error {
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/assignment.exception.ts:1]: Addressed above.

File: backend/functions/lib/use-cases/case.assignment.ts [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+import { AttorneyAssignmentResponseInterface } from '../adapters/types/case.assignment';
+import log from '../adapters/services/logger.service';
+import { applicationContextCreator } from '../adapters/utils/application-context-creator';
+import { AssignmentException } from './assignment.exception';
+
+const NAMESPACE = 'CASE-ASSIGNMENT';
+const EXISTING_ASSIGNMENT_FOUND =
+  'A trial attorney assignment already exists for this case. Cannot create another assignment on an existing case assignment.';
+export class CaseAssignment {
+  private _assignmentRepository: CaseAssignmentRepositoryInterface;
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/case.assignment.ts:14]: Addressed above - pasting the comment below - 
We decided to do that with case.attorney.assignment.ts file, not this file. Please see the related commit here -
https://github.com/US-Trustee-Program/Bankruptcy-Oversight-Support-Systems/commit/0242256321d75b93f33463bba3dd64fb2ccee880

File: backend/functions/lib/use-cases/case.assignment.ts [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  async createAssignment(
+    context: ApplicationContext,
+    caseAssignment: CaseAttorneyAssignment,
+  ): Promise<number> {
+    try {
+      return await this._assignmentRepository.createAssignment(context, caseAssignment);
+    } catch (exception) {
+      //Log the exception and handle depending.
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/case.assignment.ts:31]: Removed the comment.

File: backend/functions/lib/adapters/controllers/case.assignment.controller.ts [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+        listOfAssignments.push(assignment);
+      });
+      const assignmentService = new CaseAssignment(this.caseAssignmentRepository);
+      return assignmentService.createTrialAttorneyAssignments(
+        this.applicationContext,
+        listOfAssignments,
+      );
+    } catch (exception) {
+      log.error(applicationContextCreator(this.applicationContext), NAMESPACE, exception.message);
+      this.applicationContext.res = httpError(
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/controllers/case.assignment.controller.ts]: Done.

**Review by jayaneti1 — APPROVED** [review]

File: backend/functions/lib/adapters/controllers/case.assignment.controller.ts (hunk 30 above)
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/controllers/case.assignment.controller.ts]: Done.

File: backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts (hunk 21 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts]: Renamed to caseAssignmentLocalRepository

File: backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts (hunk 17 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/controllers/case.assignment.controller.test.ts]: The AssignmentId is created in the backend local repository for which the starting Id is 1. I updated the above test to test for toBeTruthy().

File: backend/functions/caseassignments/case.assignment.function.test.ts (hunk 20 above)
Comment (Swathi-Darbha) [reply backend/functions/caseassignments/case.assignment.function.test.ts:86]: Resolving based on our discussion to use the status code to distinguish between bad request vs internal server error. Hence keeping these tests.

File: backend/functions/lib/use-cases/case.assignment.ts [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  async isCreateValid(
+    context: ApplicationContext,
+    newAssignments: CaseAttorneyAssignment[],
+  ): Promise<boolean> {
+    const caseId = newAssignments[0].caseId;
+    const existingAssignments = await this._assignmentRepository.findAssignmentByCaseId(caseId);
+    if (existingAssignments.length === 0) return true;
+    return await this.isExactEqual(existingAssignments, newAssignments);
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/case.assignment.ts:67]: If the same exact assignment is already existing in the repository, we just return the Id(s) of those Assignment(s). Hence `isCreateValid()` returns `true` when the assignments are exactly identical. It returns `false` when any one of the assignments from the request list is different.

File: backend/functions/lib/adapters/types/trial.attorneys.assignment.request.ts [hunk 32]
```diff
@@ -0,0 +1,50 @@
...
+    this._caseId = value;
+  }
+
+  get listOfAttorneyIds(): string[] {
+    return this._listOfAttorneyIds;
+  }
+
+  set listOfAttorneyIds(value: string[]) {
+    this._listOfAttorneyIds = value;
+  }
```
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/types/trial.attorneys.assignment.request.ts:49]: Removed the accessors.

File: backend/functions/lib/adapters/types/case.assignment.request.ts (hunk 19 above)
Comment (Swathi-Darbha) [reply backend/functions/lib/adapters/types/case.assignment.request.ts:67]: Deleted the file, since this is not being called or used anywhere.

File: backend/functions/lib/adapters/gateways/case.assignment.local.repository.ts (hunk 18 above)
Comment (jamesobrooks) [reply backend/functions/lib/adapters/gateways/case.assignment.local.repository.ts:7]: Why would we not follow the style guide for all files? The style guide says not to use underscores in private member names and I showed you how you can set your IDE to handle this appropriately. I'm not saying that `_nextUnusedId` should be public, just that it should be named `nextUnusedId`.

File: backend/functions/lib/use-cases/case.assignment.ts (hunk 28 above)
[the remark this answers is not in the document]
Comment (jamesobrooks) [reply backend/functions/lib/use-cases/case.assignment.ts:14]: As in the other case, if we are following Google's style guide, we should be doing so for all files.

File: backend/functions/lib/use-cases/case.assignment.ts (hunk 31 above)
[the remark this answers is not in the document]
Comment (jamesobrooks) [reply backend/functions/lib/use-cases/case.assignment.ts:67]: I understand what we are doing. What I am wondering is why we are doing it. According to HTTP Semantics, [POST is not one of the methods that is intended to be idempotent](https://www.rfc-editor.org/rfc/rfc9110.html#name-idempotent-methods).

> A request method is considered "idempotent" if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request. Of the request methods defined by this specification, PUT, DELETE, and safe request methods are idempotent.

As the above references safe methods, here is an excerpt from that section:
> Of the request methods defined by this specification, the GET, HEAD, OPTIONS, and TRACE methods are defined to be safe.

File: backend/functions/caseassignments/case.assignment.function.test.ts (hunk 20 above)
Comment (jamesobrooks) [reply backend/functions/caseassignments/case.assignment.function.test.ts:86]: What is the expect supposed to be validating?
```
expect(appContext.res.body);
```

There's no condition.

File: backend/functions/lib/use-cases/case.assignment.ts (hunk 31 above)
[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/case.assignment.ts:67]: Lets discuss and I can show why we are doing this.

[the remark this answers is not in the document]
Comment (Swathi-Darbha) [reply backend/functions/lib/use-cases/case.assignment.ts:67]: Resolving per conversation.
